File: src/postprocess.rs
//! Depth post-processing: resize to original image size and colormap generation.

use image::RgbImage;
use log::info;

use crate::cli::ColorMode;
use crate::error::Result;

// ============================================================================
// Bilinear resize for f32 depth maps
// ============================================================================

/// Resize a depth map from model output size to the original image size.
pub fn resize_depth(
    depth_raw: &[f32],
    model_w: usize,
    model_h: usize,
    target_w: usize,
    target_h: usize,
) -> Vec<f32> {
    if model_w == target_w && model_h == target_h {
        return depth_raw.to_vec();
    }

    info!(
        "Resize depth: {}x{} -> {}x{} (bilinear)",
        model_w, model_h, target_w, target_h
    );

    let mut dst = vec![0.0f32; target_w * target_h];
    let sx = (model_w as f64) / (target_w as f64);
    let sy = (model_h as f64) / (target_h as f64);

    for y in 0..target_h {
        let sy_f = (y as f64 + 0.5) * sy - 0.5;
        let y0 = sy_f.floor() as usize;
        let y1 = (y0 + 1).min(model_h - 1);
        let fy = sy_f - y0 as f64;

        for x in 0..target_w {
            let sx_f = (x as f64 + 0.5) * sx - 0.5;
            let x0 = sx_f.floor() as usize;
            let x1 = (x0 + 1).min(model_w - 1);
            let fx = sx_f - x0 as f64;

            let v = depth_raw[y0 * model_w + x0] as f64 * (1.0 - fx) * (1.0 - fy)
                + depth_raw[y0 * model_w + x1] as f64 * fx * (1.0 - fy)
                + depth_raw[y1 * model_w + x0] as f64 * (1.0 - fx) * fy
                + depth_raw[y1 * model_w + x1] as f64 * fx * fy;
            dst[y * target_w + x] = v as f32;
        }
    }
    dst
}

// ============================================================================
// Statistics helpers
// ============================================================================

/// Linear interpolation percentile from a sorted slice.
pub fn percentile(sorted: &[f32], p: f32) -> f32 {
    if sorted.is_empty() {
        return 0.0;
    }
    let idx = p / 100.0 * (sorted.len() - 1) as f32;
    let lo = idx.floor() as usize;
    let hi = (lo + 1).min(sorted.len() - 1);
    let frac = idx - lo as f32;
    sorted[lo] * (1.0 - frac) + sorted[hi] * frac
}

/// Print depth statistics to stderr via log.
pub fn print_stats(depth: &[f32]) {
    let valid: Vec<f32> = depth.iter().copied().filter(|&d| d > 0.0).collect();
    if valid.is_empty() {
        log::warn!("No valid depth values!");
        return;
    }
    let mut sorted = valid;
    sorted.sort_by(|a, b| a.partial_cmp(b).unwrap_or(std::cmp::Ordering::Equal));
    let median = sorted[sorted.len() / 2];
    let p5 = percentile(&sorted, 5.0);
    let p95 = percentile(&sorted, 95.0);
    info!("Depth median: {:.2} m", median);
    info!("Depth 5%-95%: {:.1} ~ {:.1} m", p5, p95);
}

// ============================================================================
// JET colormap
// ============================================================================

/// OpenCV-compatible JET colormap (`cv2.COLORMAP_JET`). Returns [R, G, B].
///
/// Each channel is a trapezoid: `c = clamp(1.5 - |4t - k|, 0, 1)`
/// with k = 3, 2, 1 for R, G, B. Endpoints are dark blue / dark red.
fn jet_color(t: f32) -> [u8; 3] {
    let r = (1.5 - (4.0 * t - 3.0).abs()).clamp(0.0, 1.0);
    let g = (1.5 - (4.0 * t - 2.0).abs()).clamp(0.0, 1.0);
    let b = (1.5 - (4.0 * t - 1.0).abs()).clamp(0.0, 1.0);
    [
        (r * 255.0 + 0.5) as u8,
        (g * 255.0 + 0.5) as u8,
        (b * 255.0 + 0.5) as u8,
    ]
}

// ============================================================================
// Colormap generation
// ============================================================================

/// Convert a depth map to a JET colormap image and save it.
pub fn save_colormap(
    depth: &[f32],
    width: usize,
    height: usize,
    output_path: &str,
    mode: ColorMode,
) -> Result<()> {
    let pixels: Vec<u8> = match mode {
        ColorMode::Disparity => disparity_to_pixels(depth),
        ColorMode::Metric => metric_to_pixels(depth),
    };

    let img = RgbImage::from_raw(width as u32, height as u32, pixels)
        .ok_or_else(|| crate::error::Error::Invalid("failed to create image buffer".into()))?;
    img.save(output_path)?;
    info!("Colormap saved: {}", output_path);
    Ok(())
}

/// Disparity mode: 1/d with 2%-98% percentile clipping.
fn disparity_to_pixels(depth: &[f32]) -> Vec<u8> {
    let n = depth.len();
    let mut valid_disparity = Vec::new();
    let mut disparity = vec![0.0f32; n];

    for (i, &d) in depth.iter().enumerate() {
        if d > 0.0 {
            let disp = 1.0 / d;
            disparity[i] = disp;
            valid_disparity.push(disp);
        }
    }

    valid_disparity.sort_by(|a, b| a.partial_cmp(b).unwrap_or(std::cmp::Ordering::Equal));
    let lo = percentile(&valid_disparity, 2.0);
    let hi = percentile(&valid_disparity, 98.0);
    let range = (hi - lo).max(1e-6);

    depth
        .iter()
        .zip(disparity.iter())
        .flat_map(|(&d, &v)| {
            if d <= 0.0 {
                [0u8, 0u8, 0u8]
            } else {
                let t = ((v - lo) / range).clamp(0.0, 1.0);
                jet_color(t)
            }
        })
        .collect()
}

/// Metric mode: linear mapping over the valid depth range.
fn metric_to_pixels(depth: &[f32]) -> Vec<u8> {
    let min_d = depth.iter().copied().fold(f32::INFINITY, f32::min);
    let max_d = depth.iter().copied().fold(f32::NEG_INFINITY, f32::max);
    let range = (max_d - min_d).max(1e-8);

    depth
        .iter()
        .flat_map(|&v| {
            let t = ((v - min_d) / range).clamp(0.0, 1.0);
            jet_color(t)
        })
        .collect()
}

File: cpp/src/postprocess.cpp
#include "postprocess.hpp"

#include <algorithm>
#include <cmath>
#include <iostream>
#include <limits>
#include <vector>

#include <opencv2/imgproc.hpp>
#include <opencv2/imgcodecs.hpp>

// ============================================================================
// Bilinear resize for f32 depth maps
// ============================================================================

std::vector<float> resizeDepth(const float *depth, int model_w, int model_h,
                                int target_w, int target_h) {
    if (model_w == target_w && model_h == target_h) {
        return std::vector<float>(depth, depth + target_w * target_h);
    }


    std::vector<float> dst(target_w * target_h);
    double sx = static_cast<double>(model_w) / target_w;
    double sy = static_cast<double>(model_h) / target_h;

    for (int y = 0; y < target_h; ++y) {
        double sy_f = (y + 0.5) * sy - 0.5;
        int y0 = std::max(0, static_cast<int>(sy_f));
        int y1 = std::min(y0 + 1, model_h - 1);
        double fy = sy_f - y0;

        for (int x = 0; x < target_w; ++x) {
            double sx_f = (x + 0.5) * sx - 0.5;
            int x0 = std::max(0, static_cast<int>(sx_f));
            int x1 = std::min(x0 + 1, model_w - 1);
            double fx = sx_f - x0;

            double v = depth[y0 * model_w + x0] * (1.0 - fx) * (1.0 - fy)
                     + depth[y0 * model_w + x1] * fx * (1.0 - fy)
                     + depth[y1 * model_w + x0] * (1.0 - fx) * fy
                     + depth[y1 * model_w + x1] * fx * fy;
            dst[y * target_w + x] = static_cast<float>(v);
        }
    }
    return dst;
}

// ============================================================================
// Statistics
// ============================================================================

static float percentile(std::vector<float> sorted, float p) {
    if (sorted.empty()) return 0.0f;
    float idx = p / 100.0f * (sorted.size() - 1);
    int lo = static_cast<int>(idx);
    int hi = std::min(lo + 1, static_cast<int>(sorted.size()) - 1);
    float frac = idx - lo;
    return sorted[lo] * (1.0f - frac) + sorted[hi] * frac;
}

void printStats(const float *depth, int w, int h) {
    std::vector<float> valid;
    valid.reserve(w * h);
    for (int i = 0; i < w * h; ++i) {
        if (depth[i] > 0.0f) valid.push_back(depth[i]);
    }
    if (valid.empty()) {
        std::cout << "  No valid depth values!" << std::endl;
        return;
    }
    std::sort(valid.begin(), valid.end());
    float median = valid[valid.size() / 2];
    float p5 = percentile(valid, 5.0f);
    float p95 = percentile(valid, 95.0f);
    std::cout << "  Depth median: " << median << " m" << std::endl;
    std::cout << "  Depth 5%-95%: " << p5 << " ~ " << p95 << " m" << std::endl;
}

// ============================================================================
// JET colormap
// ============================================================================

static void jetColor(float t, uint8_t &r, uint8_t &g, uint8_t &b) {
    auto clamp01 = [](float v) { return std::max(0.0f, std::min(1.0f, v)); };
    r = static_cast<uint8_t>(clamp01(1.5f - std::abs(4.0f * t - 3.0f)) * 255.0f + 0.5f);
    g = static_cast<uint8_t>(clamp01(1.5f - std::abs(4.0f * t - 2.0f)) * 255.0f + 0.5f);
    b = static_cast<uint8_t>(clamp01(1.5f - std::abs(4.0f * t - 1.0f)) * 255.0f + 0.5f);
}

// ============================================================================
// Colormap generation
// ============================================================================

static void disparityToPixels(const float *depth, int n, std::vector<uint8_t> &pixels) {
    std::vector<float> valid_disparity;
    valid_disparity.reserve(n);
    std::vector<float> disparity(n, 0.0f);

    for (int i = 0; i < n; ++i) {
        if (depth[i] > 0.0f) {
            float d = 1.0f / depth[i];
            disparity[i] = d;
            valid_disparity.push_back(d);
        }
    }

    std::sort(valid_disparity.begin(), valid_disparity.end());
    float lo = percentile(valid_disparity, 2.0f);
    float hi = percentile(valid_disparity, 98.0f);
    float range = std::max(hi - lo, 1e-6f);

    pixels.resize(n * 3);
    for (int i = 0; i < n; ++i) {
        if (depth[i] <= 0.0f) {
            pixels[i * 3] = 0;
            pixels[i * 3 + 1] = 0;
            pixels[i * 3 + 2] = 0;
        } else {
            float t = std::max(0.0f, std::min(1.0f, (disparity[i] - lo) / range));
            uint8_t r, g, b;
            jetColor(t, r, g, b);
            pixels[i * 3] = r;
            pixels[i * 3 + 1] = g;
            pixels[i * 3 + 2] = b;
        }
    }
}

static void metricToPixels(const float *depth, int n, std::vector<uint8_t> &pixels) {
    float min_d = std::numeric_limits<float>::max();
    float max_d = std::numeric_limits<float>::lowest();
    for (int i = 0; i < n; ++i) {
        if (depth[i] > 0.0f) {
            min_d = std::min(min_d, depth[i]);
            max_d = std::max(max_d, depth[i]);
        }
    }
    if (min_d > max_d) { min_d = 0.0f; max_d = 1.0f; }
    float range = std::max(max_d - min_d, 1e-8f);

    pixels.resize(n * 3);
    for (int i = 0; i < n; ++i) {
        float t = std::max(0.0f, std::min(1.0f, (depth[i] - min_d) / range));
        jetColor(t, pixels[i * 3], pixels[i * 3 + 1], pixels[i * 3 + 2]);
    }
}

void saveColormap(const float *depth, int w, int h,
                  const std::string &path, const std::string &mode) {
    int n = w * h;
    std::vector<uint8_t> pixels;

    if (mode == "disparity") {
        disparityToPixels(depth, n, pixels);
    } else {
        metricToPixels(depth, n, pixels);
    }

    cv::Mat img(h, w, CV_8UC3, pixels.data());
    cv::cvtColor(img, img, cv::COLOR_RGB2BGR);
    cv::imwrite(path, img);
    std::cout << "Colormap saved: " << path << std::endl;
}
